### DISCUSSION TITLE
Running Nextcloud AIO with Keepalived and centralized NAS Storage for full redundancy

### DISCUSSION THREAD

**Discussion by hacketaler** [nextcloud/all-in-one#5686] [Questions] [answered]
Hi everyone,

I’ve been working on setting up a redundant Nextcloud environment using Nextcloud All-In-One (AIO) with Keepalived to manage a Virtual IP (VIP) for failover between two hosts. The overall goal is to have a highly available, redundant setup where all Nextcloud-related data is centralized on a NAS, rather than relying on local Docker volumes that are host-specific. Here's the current setup and my challenges:

**Current Setup:**
_Two Hosts Running Docker and Keepalived:_
Host 1 (Primary): Runs the Nextcloud AIO stack by default.
Host 2 (Backup): Takes over the VIP in case Host 1 goes down.

_NAS for Shared Data:_
The NAS is mounted on both hosts (volumes for ncdata and volume for the master-aio-config)
The idea is to centralize all Nextcloud-related data (including volumes for the database, Nextcloud, Redis, etc.) on the NAS to allow seamless failover.

_Keepalived Script:_
The script monitors the VIP and starts/stops services based on whether the host is in the "Master" or "Backup" state.
When the Backup takes over, it should start the AIO mastercontainer and any dependent services to take over seamlessly.

**Current Challenges:**
_Local Docker Volumes:_
By default, Nextcloud AIO creates local Docker volumes for components like the database, Redis, and Nextcloud itself (e.g., nextcloud_aio_nextcloud, nextcloud_aio_database, etc.). These volumes are not shared between the two hosts, so when the Backup host takes over, it starts with empty, newly created volumes, resulting in a “fresh” Nextcloud instance.

_Question:_
- How can I deploy/move all necessary data (e.g., Nextcloud files, database, Redis data, etc.) to the NAS and configure Nextcloud AIO to use the NAS volumes instead of local Docker volumes?
- Is it even possible to achieve full redundancy with Nextcloud AIO in this way, or is a manual setup with individual services (e.g., separate containers for database, Redis, etc.) the better approach?

_Controlling the AIO Mastercontainer from a Script:_
I want my Keepalived failover script to handle starting and stopping of all AIO-related containers. The "Stop Containers" and "Start Containers" functionality available in the AIO web GUI works perfectly to stop/start the dependent containers managed by the AIO mastercontainer. However, I haven’t found a way to replicate this behavior programmatically.

I’ve tried using something like the following command in the script:
docker exec -it nextcloud-aio-mastercontainer /var/www/docker-aio/docker-compose --project-directory=/mnt/docker-aio-config down
But there is no compose-file in the mastercontainer for all other services.

**My Current Thoughts:**
I’m starting to suspect that AIO might not be the best choice for a fully redundant, highly available setup. Maybe I should have manually set up the individual services (database, Redis, Nextcloud, etc.) so I have more granular control over storage and failover behavior. However, I wanted to check with the community first to see if there’s a way to achieve my goals using AIO.

Any advice or guidance on the following would be greatly appreciated:
- Moving all Nextcloud data (volumes) to the NAS for centralized, shared access.
- Programmatically managing the AIO mastercontainer’s dependent containers during a failover.
- Whether AIO is the right choice for this use case, or if I should consider a manual setup instead.

Thanks in advance for your help!

**Comment by szaimen** [comment, marked as answer]
Hi, I would recommend you to use this: https://github.com/nextcloud/all-in-one/tree/main/manual-install. See https://github.com/nextcloud/all-in-one/blob/main/manual-install/latest.yml
